src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/SystemClassWithMembersAndTypesRecord.cs

```diff
@@ -75,29 +75,30 @@ ulong value when TypeNameMatches(typeof(UIntPtr)) => Create(new UIntPtr(value)),
                     _ => this
                 };
             }
-            else if (HasMember("_ticks") && MemberValues[0] is long ticks && TypeNameMatches(typeof(TimeSpan)))
+            else if (HasMember("_ticks") && GetRawValue("_ticks") is long ticks && TypeNameMatches(typeof(TimeSpan)))
             {
                 return Create(new TimeSpan(ticks));
             }
         }
         else if (MemberValues.Count == 2
```

Comment:
- This assumes that the BF format is set is stone and that new key/value pairs won't be added in future. Is it safe assumption to make?

- If somebody constructs malicious payload with extra TimeSpan, DateTime or Guid fields or with fields of unexpected type, this pattern match won't kick in, there won't be any exception thrown and we return the raw data. Is it the desired behavior for Nrbf reader? (As far as I can tell, the reader tends to throw on anything unexpected or invalid instead of accepting it silently.)

Reply:
These are great questions.

`BinaryFormatter` can represent same primitive value using different record types based on the context.
In this case, when `DateTime` is the root object it is expressed as `SystemClassWithMembersAndTypesRecord` which is just a type name + key/value dictionary. In other cases, it can be represented as `MemberPrimitiveTypedRecord<T>` (or a raw 8 bytes).

My goal was to hide this from the end users and always map it to `PrimitiveTypeRecord<T>` so users don't need to become experts in this area.

> This assumes that the BF format is set is stone and that new key/value pairs won't be added in future.

If we ever extend the binary representation of given types, we may need to handle the versioning here.


> If somebody constructs malicious payload with extra TimeSpan, DateTime or Guid fields or with fields of unexpected type, this pattern match won't kick in, there won't be any exception thrown and we return the raw data. Is it the desired behavior for Nrbf reader?

It's allowed to create a type that is called `System.DateTime` and has different layout, in such cases we are going to return a `ClassRecord` and the users will need to handle it.


```csharp
SerializationRecord rootObject = NrbfDecoder.Decode(payload);
if (rootObject is PrimitiveTypeRecord<DateTime> primitiveRecord)
{
    // DateTime
}
else if (rootObject is ClassRecord classRecord)
{
    // something else
}
```

@jkotas this is just the way I see it, please let me know if something is not clear or some other changes are needed

Reply:
> My goal was to hide this from the end users and always map it to PrimitiveTypeRecord<T>

You are not *always* mapping it to `PrimitiveTypeRecord<T>`.

You are only mapping it to `PrimitiveTypeRecord<T>` if the input has specific shape. You are not mapping it for all possible valid input shapes. For example, if the payload was produced by .NET Framework 1.x (I am sure there are a bunch of such payloads still alive in the wild), it will be missing `dateData` field and it is not going to mapped. However, the classic BF deserializer is going to handle it just fine. If somebody runs into this case, they will have to do double the work: They will need to handle both mapped and the non-mapped cases.

In general, I would expect the behavior to be either:
- 100% compatible with classic BF deserializer
- Exception to be thrown